README.md

```diff
@@ -112,52 +136,111 @@ func main() {
+VM. The following snippet from [the previous section](#use) demonstrates
+this helper.
+
+```go
+import github.com/Azure/go-autorest/autorest/azure/auth
+
+// create a VirtualNetworks client
+vnetClient := network.NewVirtualNetworksClient("<subscriptionID>")
+
+// create an authorizer from env vars or Azure Managed Service Idenity
+authorizer, err := auth.NewAuthorizerFromEnvironment()
+if err == nil {
+    vnetClient.Authorizer = authorizer
+}
+
+// call the VirtualNetworks CreateOrUpdate API
+vnetClient.CreateOrUpdate(context.Background(),
+// ...
+```
+
+The following environment variables help determine authentication configuration:
+
+- `AZURE_ENVIRONMENT`: Specifies the Azure Environment to use. If not set, it
+  defaults to `AzurePublicCloud`. Not applicable to authentication with Managed
+  Service Identity (MSI).
+- `AZURE_AD_RESOURCE`: Specifies the AAD resource ID to use. If not set, it
+  defaults to `ResourceManagerEndpoint` for operations with Azure Resource
+  Manager.
+
+## Authentication Details
+
+The previous is the first and most recommended of several authentication
+options offered by the SDK because it allows seamless use of both service
+principals and [Azure Managed Service Identity][]. Other options are listed
+below.
+
+> Note: If you need to create a new service principal, run `az ad sp
+> create-for-rbac -n "<app_name>"` in the
+> [azure-cli](https://github.com/Azure/azure-cli). See [these
+> docs](https://docs.microsoft.com/cli/azure/create-an-azure-service-principal-azure-cli?view=azure-cli-latest)
+> for more info. Copy the new principal's ID, secret, and tenant ID for use in
+> your app, or consider the `--sdk-auth` parameter for serialized output.
+
+[Azure Managed Service Identity]: https://docs.microsoft.com/en-us/azure/active-directory/msi-overview
+
+* The `auth.NewAuthorizerFromEnvironment()` described above creates an authorizer
+from the first available of the following configuration:
+
+    1. **Client Credentials**: Azure AD Application ID and Secret.
+
+        - `AZURE_TENANT_ID`: Specifies the Tenant to which to authenticate.
+        - `AZURE_CLIENT_ID`: Specifies the app client ID to use.
+        - `AZURE_CLIENT_SECRET`: Specifies the app secret to use.
+
+    2. **Client Certificate**: Azure AD Application ID and X.509 Certificate.
+
+        - `AZURE_TENANT_ID`: Specifies the Tenant to which to authenticate.
+        - `AZURE_CLIENT_ID`: Specifies the app client ID to use.
+        - `AZURE_CERTIFICATE_PATH`: Specifies the certificate Path to use.
+        - `AZURE_CERTIFICATE_PASSWORD`: Specifies the certificate password to use.
+
+    3. **Resource Owner**: Azure AD User and Password. This grant type is *not
+       recommended*, use client credentials (service principals) instead.
+
+        - `AZURE_TENANT_ID`: Specifies the Tenant to which to authenticate.
+        - `AZURE_CLIENT_ID`: Specifies the app client ID to use.
```

Comment:
I really really wish that we had two separate environment variables for this. It doesn't make any sense to me that we've overloaded it.

Reply:
with what are we overloading it ? doesn't always refer to the app id to use ?

---
In reply to: [195175985](https://github.com/Azure/azure-sdk-for-go/pull/1911#discussion_r195175985) [](ancestors = 195175985)

Reply:
It means the same thing here as in the other flows - the ID of the app requesting access "on behalf of" the resource owner.

Reply:
yea either way this thing will always be confusing for users and not only. in the AAD Graph API there   there is even no client id field for neither the SP object or App Object. Not sure why they went with calling it client id instead of app id